webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,239 @@
+  id?: string;
+}
+
+interface SubscriptionWithCache extends Subscription {
+  keyCache: KeyCache;
+}
+
+type SubscriptionGroup = Partial<Record<Streamable, Subscription>>;
+
+export class Stream {
+  readonly #wsUrl: string;
+  #ws?: WebSocket = undefined;
+  #retries: number = 0;
+  #timeout: Dayjs = dayjs();
+  #numSyncs: number = 0;
+  #closedByClient: boolean = false;
+
+  #subs: Array<SubscriptionGroup> = [];
+  #curSub?: Partial<Record<Streamable, SubscriptionWithCache>>;
+
+  // syncSent updates when msg sent to the stream
+  #syncSent?: string = undefined;
+  // syncStarted updates when recieving msg of {sync_id, complated: false}
+  #syncStarted?: string = undefined;
+  // syncStarted updates when recieving msg of {sync_id, complated: true}
+  #syncComplete?: string = undefined;
+  // List of messages recieved from server
+  #pendingMsg: Array<Record<string, StreamContent>> = [];
+
+  //callbacks
+  #onUpsert: (m: Record<string, StreamContent>) => void;
+  #onDelete: (s: Streamable, a: Array<number>) => void;
+  #isLoaded?: (ids: Array<string>) => void;
+
+  constructor(
+    wsUrl: string,
+    onUpsert: (m: Record<string, StreamContent>) => void,
+    onDelete: (s: Streamable, a: Array<number>) => void,
+    isLoaded?: (ids: Array<string>) => void,
+  ) {
+    this.#wsUrl = wsUrl;
+    this.#onUpsert = onUpsert;
+    this.#onDelete = onDelete;
+    this.#isLoaded = isLoaded;
+    this.#advance();
+  }
+
+  #connect(): WebSocket {
+    const ws = new WebSocket(this.#wsUrl);
+    ws.onopen = () => {
+      logger.info('Streaming websocket opened!');
+      this.#retries = 0;
+      this.#advance();
+    };
+    ws.onerror = (err) => {
+      // No need to do anything else becauses onerror will trigger onclose
+      logger.error('Streaming websocket errored: ', err);
+    };
+
+    ws.onclose = () => {
+      this.#syncSent = undefined;
+      const backoff = backoffs[this.#retries];
+      if (backoff === undefined) {
+        throw new Error('Websocket cannot reconnect!');
+      }
+      this.#timeout = dayjs().add(backoff, 'second');
+      this.#retries += 1;
+      logger.info(`#${this.#retries} of retries: in ${backoff}s`);
+      setTimeout(() => this.#advance(), backoff * 1000);
+    };
+
+    ws.onmessage = (event) => {
+      const msg = JSON.parse(event.data) as Record<string, StreamContent>;
+      this.#pendingMsg.push(msg);
+      this.#advance();
+    };
+
+    return ws;
+  }
+
+  #shouldSkip(newSub: SubscriptionGroup): boolean {
+    if (!this.#curSub) return false;
+    let same = true;
+    forEach(newSub, (val, k) => {
+      if (!this.#curSub?.[k as Streamable]?.spec.equals(val?.spec)) same = false;
+    });
+    return same;
+  }
+
+  #sendSpec(newSub: SubscriptionGroup): void {
+    // Skip current sub and move to next only if sub has already been sent
+    if (this.#shouldSkip(newSub) && this.#syncSent) {
+      this.#advance();
+      return;
+    }
```

Comment:
I'm not confident that this shouldSkip logic is 100% correct.

-  What if the caller submits a spec.  Suppose it is a duplicate of the last spec but the client doesn't know that yet.  They might be expecting to see certain callbacks, but as written it's not well-defined if they should expect callbacks from the Stream or not.  They'd have to replicate this deduplication logic in order to know how to act.

- What if you skip sending a duplicate spec, and there is third spec in the queue.  You decide here not to send the second spec because shouldSkip returns true.  But when would you send that third spec?

Reply:
> Not sure I get the question, when we decide to skip, we advance, and that's when we send the third spec

Ah, I misread.  Let me read again.